OpenRA.Mods.Common/HitShapes/Capsule.cs

```diff
@@ -73,22 +84,44 @@ public WDist DistanceFromEdge(WVec v)
 
 		public WDist DistanceFromEdge(WPos pos, Actor actor)
 		{
-			return DistanceFromEdge((pos - actor.CenterPosition).Rotate(-actor.Orientation));
+			var actorPos = actor.CenterPosition;
+
+			if (pos.Z <= actorPos.Z + VerticalTopOffset && pos.Z >= actorPos.Z + VerticalBottomOffset)
+				return DistanceFromEdge((pos - new WPos(actorPos.X, actorPos.Y, pos.Z)).Rotate(-actor.Orientation));
+
+			if (pos.Z > actorPos.Z + VerticalTopOffset)
+				return DistanceFromEdge((pos - (actorPos + new WVec(0, 0, VerticalTopOffset))).Rotate(-actor.Orientation));
+
+			if (pos.Z < actorPos.Z + VerticalBottomOffset)
+				return DistanceFromEdge((pos - (actorPos + new WVec(0, 0, VerticalBottomOffset))).Rotate(-actor.Orientation));
+
+			return DistanceFromEdge((pos - actorPos).Rotate(-actor.Orientation));
```

Comment:
Correct me if I'm wrong, but won't we ever reach this (because of the checks above)?

As soon as one of the conditions from the first check is false, one of the other two checks will be true.